sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessor.cs

```diff
@@ -689,27 +722,51 @@ public virtual async Task StopProcessingAsync(CancellationToken cancellationToke
         private async Task RunReceiveTaskAsync(
             CancellationToken cancellationToken)
         {
-            List<Task> tasks = new List<Task>(_maxConcurrentCalls + _receiverManagers.Count);
             try
             {
                 while (!cancellationToken.IsCancellationRequested && !Connection.IsClosed)
                 {
-                    foreach (ReceiverManager receiverManager in _receiverManagers)
+                    ReceiverManager[] receiverManagers = null;
+                    lock (_maxConcurrencySyncLock)
+                    {
+                        // create a new list of receiver managers as the underlying list can change while running
+                        receiverManagers = _receiverManagers.ToArray();
```

Comment:
Instead of creating a new array per iteration create it in the UpdateConcurrency